File: rust/src/solver.rs
use crate::check_validity::check_question_against_solution;
use crate::types::*;

pub fn solve(
    fp: &FlatPuzzle,
    fixed: Option<&[Option<Answer>; MAX_N]>,
    max_solutions: usize,
) -> Vec<[Answer; MAX_N]> {
    let n = fp.n;
    let default_fixed = [None; MAX_N];
    let fixed = fixed.unwrap_or(&default_fixed);
    let mut solutions: Vec<[Answer; MAX_N]> = Vec::new();
    let mut current = [None::<Answer>; MAX_N];
    let order = compute_search_order(fp);
    let all_bits: u16 = (1u16 << n) - 1;
    let mut assigned_bits: u16 = 0;

    let range_masks = compute_range_masks(fp);

    search(
        fp,
        fixed,
        &mut solutions,
        &mut current,
        &order,
        all_bits,
        &mut assigned_bits,
        &range_masks,
        0,
        max_solutions,
    );

    solutions
}

fn compute_search_order(fp: &FlatPuzzle) -> [u8; MAX_N] {
    let n = fp.n;
    let mut ref_count = [0u8; MAX_N];
    for i in 0..n {
        if let QuestionType::AnswerOf { question_index } = fp.question_types[i] {
            ref_count[question_index as usize] += 1;
        }
    }

    let mut indices: [u8; MAX_N] = std::array::from_fn(|i| i as u8);
    indices[..n].sort_by(|&a, &b| {
        let a = a as usize;
        let b = b as usize;
        ref_count[b].cmp(&ref_count[a]).then_with(|| {
            let a_global = fp.question_types[a].is_solver_global() as u8;
            let b_global = fp.question_types[b].is_solver_global() as u8;
            a_global.cmp(&b_global)
        })
    });

    indices
}

fn compute_range_masks(fp: &FlatPuzzle) -> [u16; MAX_N] {
    let n = fp.n;
    let mut masks = [0u16; MAX_N];
    for i in 0..n {
        masks[i] = match fp.question_types[i] {
            QuestionType::NextSame => {
                let mut m = 0u16;
                for j in (i + 1)..n {
                    m |= 1 << j;
                }
                m
            }
            QuestionType::ClosestAfter { after_index, .. }
            | QuestionType::CountAnswerAfter { after_index, .. } => {
                let mut m = 0u16;
                for j in (after_index as usize + 1)..n {
                    m |= 1 << j;
                }
                m
            }
            QuestionType::ClosestBefore { before_index, .. }
            | QuestionType::CountAnswerBefore { before_index, .. } => {
                let mut m = 0u16;
                for j in 0..before_index as usize {
                    m |= 1 << j;
                }
                m
            }
            _ => 0,
        };
    }
    masks
}

#[allow(clippy::too_many_arguments)]
fn search(
    fp: &FlatPuzzle,
    fixed: &[Option<Answer>; MAX_N],
    solutions: &mut Vec<[Answer; MAX_N]>,
    current: &mut [Option<Answer>; MAX_N],
    order: &[u8; MAX_N],
    all_bits: u16,
    assigned_bits: &mut u16,
    range_masks: &[u16; MAX_N],
    depth: usize,
    max_solutions: usize,
) {
    let n = fp.n;
    if solutions.len() >= max_solutions {
        return;
    }

    if depth == n {
        let mut valid = true;
        for i in 0..n {
            if !check_question_against_solution(fp, i, current[i].unwrap(), current) {
                valid = false;
                break;
            }
        }
        if valid {
            let mut copy = [Answer::A; MAX_N];
            for i in 0..n {
                copy[i] = current[i].unwrap();
            }
            solutions.push(copy);
        }
        return;
    }

    let qi = order[depth] as usize;
    let bit = 1u16 << qi;

    if let Some(letter) = fixed[qi] {
        current[qi] = Some(letter);
        *assigned_bits |= bit;
        if !has_contradiction(fp, current, n, qi, *assigned_bits, all_bits, range_masks) {
            search(
                fp,
                fixed,
                solutions,
                current,
                order,
                all_bits,
                assigned_bits,
                range_masks,
                depth + 1,
                max_solutions,
            );
        }
        current[qi] = None;
        *assigned_bits &= !bit;
        return;
    }

    for &letter in &LETTERS {
        current[qi] = Some(letter);
        *assigned_bits |= bit;
        if !has_contradiction(fp, current, n, qi, *assigned_bits, all_bits, range_masks) {
            search(
                fp,
                fixed,
                solutions,
                current,
                order,
                all_bits,
                assigned_bits,
                range_masks,
                depth + 1,
                max_solutions,
            );
            if solutions.len() >= max_solutions {
                current[qi] = None;
                *assigned_bits &= !bit;
                return;
            }
        }
    }
    current[qi] = None;
    *assigned_bits &= !bit;
}

fn has_contradiction(
    fp: &FlatPuzzle,
    answers: &[Option<Answer>; MAX_N],
    n: usize,
    just_assigned: usize,
    assigned: u16,
    all_bits: u16,
    range_masks: &[u16; MAX_N],
) -> bool {
    let all_answered = assigned == all_bits;

    if all_answered {
        for i in 0..n {
            if !check_question_against_solution(fp, i, answers[i].unwrap(), answers) {
                return true;
            }
        }
        return false;
    }

    let affected = &fp.affected_by[just_assigned];
    for k in 0..affected.len as usize {
        let i = affected.data[k] as usize;
        if answers[i].is_none() {
            continue;
        }
        if check_rule(fp, answers, n, i, all_answered, assigned, range_masks) {
            return true;
        }
    }

    let globals = &fp.global_indices;
    for k in 0..globals.len as usize {
        let i = globals.data[k] as usize;
        if answers[i].is_none() || i == just_assigned {
            continue;
        }
        if check_rule(fp, answers, n, i, all_answered, assigned, range_masks) {
            return true;
        }
    }

    false
}

fn check_rule(
    fp: &FlatPuzzle,
    answers: &[Option<Answer>; MAX_N],
    n: usize,
    i: usize,
    all_answered: bool,
    assigned: u16,
    range_masks: &[u16; MAX_N],
) -> bool {
    let t = &fp.question_types[i];
    let answer_i = answers[i].unwrap();

    if (all_answered || can_fully_evaluate_local(t, assigned, range_masks, i))
        && !check_question_against_solution(fp, i, answer_i, answers)
    {
        return true;
    }

    // Forward checking for counting types
    match *t {
        QuestionType::CountAnswer { answer }
        | QuestionType::CountAnswerBefore { answer, .. }
        | QuestionType::CountAnswerAfter { answer, .. } => {
            let opt_val = fp.option_nums[i][answer_i.idx()];
            if opt_val == NAN_VAL {
                return false;
            }

            let (range_start, range_end) = match *t {
                QuestionType::CountAnswer { .. } => (0, n),
                QuestionType::CountAnswerBefore { before_index, .. } => (0, before_index as usize),
                QuestionType::CountAnswerAfter { after_index, .. } => (after_index as usize + 1, n),
                _ => unreachable!(),
            };

            let mut count: i16 = 0;
            let mut remaining: i16 = 0;
            for j in range_start..range_end {
                if answers[j] == Some(answer) {
                    count += 1;
                } else if answers[j].is_none() {
                    remaining += 1;
                }
            }
            if count > opt_val || count + remaining < opt_val {
                return true;
            }
        }
        QuestionType::CountVowel | QuestionType::CountConsonant => {
            let opt_val = fp.option_nums[i][answer_i.idx()];
            if opt_val == NAN_VAL {
                return false;
            }
            let is_vowel = matches!(*t, QuestionType::CountVowel);
            let mut count: i16 = 0;
            let mut remaining: i16 = 0;
            for j in 0..n {
                if answers[j].is_none() {
                    remaining += 1;
                } else if let Some(a) = answers[j]
                    && is_vowel == a.is_vowel()
                {
                    count += 1;
                }
            }
            if count > opt_val || count + remaining < opt_val {
                return true;
            }
        }
        _ => {}
    }

    false
}

fn can_fully_evaluate_local(
    t: &QuestionType,
    assigned: u16,
    range_masks: &[u16; MAX_N],
    qi: usize,
) -> bool {
    match *t {
        QuestionType::AnswerIsSelf => true,
        QuestionType::PrevSame => {
            let mut mask = 0u16;
            for j in 0..qi {
                mask |= 1 << j;
            }
            (assigned & mask) == mask
        }
        QuestionType::AnswerOf { question_index } => (assigned & (1 << question_index)) != 0,
        QuestionType::LetterDist { question_index } => (assigned & (1 << question_index)) != 0,
        QuestionType::NextSame
        | QuestionType::ClosestAfter { .. }
        | QuestionType::ClosestBefore { .. }
        | QuestionType::CountAnswerBefore { .. }
        | QuestionType::CountAnswerAfter { .. } => {
            let mask = range_masks[qi];
            (assigned & mask) == mask
        }
        _ => false,
    }
}

File: src/generator/solver.ts
import type {
  AnswerLetter,
  Puzzle,
  FlatPuzzle,
  FlatQuestion,
  QuestionTypeId,
} from "../engine/types.ts";
import {
  LETTERS,
  letterIdx,
  flattenPuzzle,
  RT_COUNT_ANSWER,
  RT_COUNT_ANSWER_BEFORE,
  RT_COUNT_ANSWER_AFTER,
  RT_COUNT_VOWEL,
  RT_COUNT_CONSONANT,
  RT_MOST_COMMON_COUNT,
  RT_CLOSEST_AFTER,
  RT_CLOSEST_BEFORE,
  RT_PREV_SAME,
  RT_NEXT_SAME,
  RT_ONLY_SAME,
  RT_CONSEC_IDENT,
  RT_ONLY_ODD,
  RT_ANSWER_OF,
  RT_LEAST_COMMON,
  RT_MOST_COMMON,
  RT_UNIQUE,
  RT_EQUAL_COUNT,
  RT_SELF,
  RT_LETTER_DIST,
  RT_TRUE_STMT,
} from "../engine/types.ts";
import { checkQuestionAgainstSolution as evaluate } from "../engine/check-validity.ts";

export function solve(
  puzzle: Puzzle,
  fixedAnswers?: (AnswerLetter | null)[],
  maxSolutions = 2,
): AnswerLetter[][] {
  return solveFp(flattenPuzzle(puzzle), fixedAnswers, maxSolutions);
}

const SOLVER_GLOBAL_IDS: Set<QuestionTypeId> = new Set([
  RT_COUNT_ANSWER,
  RT_COUNT_VOWEL,
  RT_COUNT_CONSONANT,
  RT_LEAST_COMMON,
  RT_MOST_COMMON,
  RT_MOST_COMMON_COUNT,
  RT_UNIQUE,
  RT_EQUAL_COUNT,
  RT_TRUE_STMT,
  RT_ONLY_SAME,
  RT_CONSEC_IDENT,
  RT_ONLY_ODD,
]);

function computeSearchOrder(fp: FlatPuzzle): number[] {
  const n = fp.n;

  // Count how many answer_of_question rules point to each question
  const refCount = new Array<number>(n).fill(0);
  for (const q of fp.questions) {
    if (q.t === RT_ANSWER_OF && q.questionIndex >= 0) {
      refCount[q.questionIndex]++;
    }
  }

  const indices = Array.from({ length: n }, (_, i) => i);

  indices.sort((a, b) => {
    // Most-referenced questions first (they unlock answer_of_question chains)
    if (refCount[b] !== refCount[a]) return refCount[b] - refCount[a];
    // Non-global rules before global (global needs all answers)
    const aGlobal = SOLVER_GLOBAL_IDS.has(fp.questions[a].t) ? 1 : 0;
    const bGlobal = SOLVER_GLOBAL_IDS.has(fp.questions[b].t) ? 1 : 0;
    return aGlobal - bGlobal;
  });

  return indices;
}

function solveFp(
  fp: FlatPuzzle,
  fixedAnswers?: (AnswerLetter | null)[],
  maxSolutions = 2,
): AnswerLetter[][] {
  const n = fp.n;
  const fixed = fixedAnswers ?? new Array<AnswerLetter | null>(n).fill(null);
  const solutions: AnswerLetter[][] = [];
  const current = new Array<AnswerLetter | null>(n).fill(null);
  const order = computeSearchOrder(fp);
  const allBits = (1 << n) - 1;
  let assignedBits = 0;

  // Pre-compute bitmasks for range checks in canFullyEvaluateLocal
  const rangeMasks: number[] = new Array(n);
  for (let i = 0; i < n; i++) {
    const q = fp.questions[i];
    if (q.t === RT_NEXT_SAME) {
      let m = 0;
      for (let j = i + 1; j < n; j++) m |= 1 << j;
      rangeMasks[i] = m;
    } else if (q.t === RT_CLOSEST_AFTER) {
      let m = 0;
      for (let j = q.afterIndex + 1; j < n; j++) m |= 1 << j;
      rangeMasks[i] = m;
    } else if (q.t === RT_CLOSEST_BEFORE || q.t === RT_COUNT_ANSWER_BEFORE) {
      let m = 0;
      for (let j = 0; j < q.beforeIndex; j++) m |= 1 << j;
      rangeMasks[i] = m;
    } else if (q.t === RT_COUNT_ANSWER_AFTER) {
      let m = 0;
      for (let j = q.afterIndex + 1; j < n; j++) m |= 1 << j;
      rangeMasks[i] = m;
    } else {
      rangeMasks[i] = 0;
    }
  }

  function search(depth: number) {
    if (solutions.length >= maxSolutions) return;

    if (depth === n) {
      let valid = true;
      for (let i = 0; i < n; i++) {
        if (!evaluate(fp, i, current[i]!, current)) {
          valid = false;
          break;
        }
      }
      if (valid) {
        const copy: AnswerLetter[] = [];
        for (let i = 0; i < n; i++) copy.push(current[i]!);
        solutions.push(copy);
      }
      return;
    }

    const qi = order[depth];
    const bit = 1 << qi;

    if (fixed[qi] != null) {
      current[qi] = fixed[qi];
      assignedBits |= bit;
      if (
        !hasContradiction(fp, current, n, qi, assignedBits, allBits, rangeMasks)
      ) {
        search(depth + 1);
      }
      current[qi] = null;
      assignedBits &= ~bit;
      return;
    }

    for (const letter of LETTERS) {
      current[qi] = letter;
      assignedBits |= bit;
      if (
        !hasContradiction(fp, current, n, qi, assignedBits, allBits, rangeMasks)
      ) {
        search(depth + 1);
        if (solutions.length >= maxSolutions) {
          current[qi] = null;
          assignedBits &= ~bit;
          return;
        }
      }
    }
    current[qi] = null;
    assignedBits &= ~bit;
  }

  search(0);
  return solutions;
}

function hasContradiction(
  fp: FlatPuzzle,
  answers: (AnswerLetter | null)[],
  n: number,
  justAssigned: number,
  assigned: number,
  allBits: number,
  rangeMasks: number[],
): boolean {
  const allAnswered = assigned === allBits;

  // When all answered, full check (correctness guarantee)
  if (allAnswered) {
    for (let i = 0; i < n; i++) {
      if (!evaluate(fp, i, answers[i]!, answers)) return true;
    }
    return false;
  }

  // Incremental: check questions affected by the just-assigned question
  const affected = fp.affectedBy[justAssigned];
  for (let k = 0; k < affected.length; k++) {
    const i = affected[k];
    if (answers[i] == null) continue;
    if (checkRule(fp, answers, n, i, allAnswered, assigned, rangeMasks))
      return true;
  }

  // Check global rules for forward-checking bounds
  const globals = fp.globalIndices;
  for (let k = 0; k < globals.length; k++) {
    const i = globals[k];
    if (answers[i] == null) continue;
    if (i === justAssigned) continue;
    if (checkRule(fp, answers, n, i, allAnswered, assigned, rangeMasks))
      return true;
  }

  return false;
}

function checkRule(
  fp: FlatPuzzle,
  answers: (AnswerLetter | null)[],
  n: number,
  i: number,
  allAnswered: boolean,
  assigned: number,
  rangeMasks: number[],
): boolean {
  const q = fp.questions[i];

  if (
    allAnswered ||
    canFullyEvaluateLocal(q, answers, assigned, rangeMasks, i)
  ) {
    if (!evaluate(fp, i, answers[i]!, answers)) return true;
  }

  // Forward checking for counting rules
  if (
    q.t === RT_COUNT_ANSWER ||
    q.t === RT_COUNT_ANSWER_BEFORE ||
    q.t === RT_COUNT_ANSWER_AFTER
  ) {
    const optVal = fp.optionValues[i][letterIdx(answers[i]!)];
    if (optVal == null) return false;

    let rangeStart: number;
    let rangeEnd: number;

    if (q.t === RT_COUNT_ANSWER) {
      rangeStart = 0;
      rangeEnd = n;
    } else if (q.t === RT_COUNT_ANSWER_BEFORE) {
      rangeStart = 0;
      rangeEnd = q.beforeIndex;
    } else {
      rangeStart = q.afterIndex + 1;
      rangeEnd = n;
    }

    let count = 0;
    let remaining = 0;
    for (let j = rangeStart; j < rangeEnd; j++) {
      if (answers[j] === q.answer) count++;
      else if (answers[j] == null) remaining++;
    }
    if (count > optVal || count + remaining < optVal) return true;
  }

  if (q.t === RT_COUNT_VOWEL || q.t === RT_COUNT_CONSONANT) {
    const optVal = fp.optionValues[i][letterIdx(answers[i]!)];
    if (optVal == null) return false;
    const isVowel = q.t === RT_COUNT_VOWEL;
    let count = 0;
    let remaining = 0;
    for (let j = 0; j < n; j++) {
      if (answers[j] == null) {
        remaining++;
      } else if (
        isVowel
          ? answers[j] === "A" || answers[j] === "E"
          : answers[j] !== "A" && answers[j] !== "E"
      ) {
        count++;
      }
    }
    if (count > optVal || count + remaining < optVal) return true;
  }

  return false;
}

// Lightweight canFullyEvaluate for non-global rules only
function canFullyEvaluateLocal(
  q: FlatQuestion,
  _answers: (AnswerLetter | null)[],
  assigned: number,
  rangeMasks: number[],
  questionIdx: number,
): boolean {
  switch (q.t) {
    case RT_SELF:
      return true;
    case RT_PREV_SAME: {
      let mask = 0;
      for (let j = 0; j < questionIdx; j++) mask |= 1 << j;
      return (assigned & mask) === mask;
    }
    case RT_ANSWER_OF:
      return (assigned & (1 << q.questionIndex)) !== 0;
    case RT_LETTER_DIST:
      return (assigned & (1 << q.questionIndex)) !== 0;
    case RT_NEXT_SAME:
    case RT_CLOSEST_AFTER:
    case RT_CLOSEST_BEFORE:
    case RT_COUNT_ANSWER_BEFORE:
    case RT_COUNT_ANSWER_AFTER: {
      const mask = rangeMasks[questionIdx];
      return (assigned & mask) === mask;
    }
    default:
      return false;
  }
}
